main_test.go

```diff
@@ -0,0 +1,714 @@
+			}
+			userToPassword := map[string]string{}
+			for rowIdx, expected := range tc.PasswordManagerOut {
+				got := pmRows[rowIdx+1]
+				if expected.Username != got[0] {
+					t.Fatalf("%s Row %d: expected Username=%s but got Username=%s",
+						sheetNamePasswordManager, rowIdx+1, expected.Username, got[0])
+				}
+				// If the password has changed, we can't predict what it
+				// will be, so just check that it matches what was sent to
+				// the server.
+				newPassword, ok := handler.UserToNewPassword[expected.Username]
+				if ok {
+					if expected.Password != newPasswordMarker {
+						t.Fatalf("%s's password was changed unexpectedly", expected.Username)
+					} else if newPassword != got[1] {
+						t.Fatalf("%s Row %d: new password recorded as %q but sent to server as %q",
+							sheetNamePasswordManager, rowIdx+1, got[1], newPassword)
+					}
+					delete(handler.UserToNewPassword, expected.Username)
+				} else if expected.Password == newPasswordMarker {
+					t.Fatalf("Expected a password change for %s but the server did not get one", expected.Username)
+				}
+				if expected.Previous != got[2] {
+					t.Fatalf("%s Row %d: expected Previous=%s but got Previous=%s",
+						sheetNamePasswordManager, rowIdx+1, expected.Previous, got[2])
+				}
+				expectedTimestamp := now.Add(expected.Timestamp)
+				gotTimestamp, err := time.Parse(time.UnixDate, got[3])
+				if err != nil {
+					t.Fatalf("%s Row %d: error parsing date %q: %s", sheetNamePasswordManager, rowIdx+1, got[3], err)
+				}
+				diff := expectedTimestamp.Sub(gotTimestamp)
+				maxDiff := time.Hour
+				if diff < -maxDiff || diff > maxDiff {
+					t.Fatalf("%s Row %d: expected Timestamp~=%s but got Timestamp=%s",
+						sheetNamePasswordManager, rowIdx+1, expectedTimestamp, gotTimestamp)
+				}
+				userToPassword[got[0]] = got[1]
+			}
+			if len(handler.UserToNewPassword) != 0 {
+				t.Fatalf("Passwords were updated for users not in the manager: %v", handler.UserToNewPassword)
+			}
+
+			macFinUsers := map[string]struct{}{}
+			for _, row := range tc.MACFinIn {
+				if row.Username != "" {
+					macFinUsers[row.Username] = struct{}{}
+				}
+			}
+			macFinRows, err := f.GetRows(sheetNameMACFin)
+			if err != nil {
+				t.Fatalf("Error getting MACFin rows: %s", err)
+			}
+			for idx, row := range macFinRows[1:] {
+				if len(row) < 6 {
+					continue // short, invalid row
+				}
+				username := row[4]
+				password := row[5]
+				expectedPassword, ok := userToPassword[username]
+				if !ok {
+					t.Fatalf("%s row %d: user %q is not in password manager",
+						sheetNameMACFin, idx+1, username)
+				} else if expectedPassword != password {
+					t.Fatalf("%s row %d: Expected password %q but got %q",
+						sheetNameMACFin, idx+1, expectedPassword, password)
+				}
+				delete(macFinUsers, username)
+			}
+			if len(macFinUsers) != 0 {
+				t.Fatalf("Users are now missing from the MACFin sheet: %v", macFinUsers)
```

Comment:
Should this be "Users are now missing from the password manager sheet?"

Reply:
No, that would be found when we check `PasswordManagerOut`